Arrt/Widgets/FlowLayout.cpp

```diff
@@ -0,0 +1,284 @@
+        }
+        w += item->sizeHint().width();
+        h = qMax(item->sizeHint().height(), h);
+    }
+    h += 2 * margin();
+    return QSize(0, h);
+}
+
+QSize FlowLayout::minimumSize() const
+{
+    QSize size;
+    for (auto iter = m_items.begin(); iter != m_items.end(); ++iter)
+    {
+        QLayoutItem* item = *iter;
+        if (item->widget() && !item->widget()->isVisibleTo(parentWidget()))
+        {
+            continue;
+        }
+        size = size.expandedTo(item->minimumSize());
+    }
+
+    size += QSize(2 * margin(), 2 * margin());
+    return size;
+}
+
+int FlowLayout::doLayout(const QRect& rect, bool testOnly) const
+{
+    int left, top, right, bottom;
+    getContentsMargins(&left, &top, &right, &bottom);
+    QRect effectiveRect = rect.adjusted(left, top, -right, -bottom);
+    int x = effectiveRect.x();
+    int y = effectiveRect.y();
+    int lineHeight = 0;
+
+    QList<QLayoutItem*> visibleItems;
+
+    // for layouting we may only take the visible items into consideration
+    // it becomes much easier to do this with a copy of the array, containing only the visible items
+    for (auto iter = m_items.begin(); iter != m_items.end(); ++iter)
+    {
+        QLayoutItem* item = *iter;
+        if (item->widget() && item->widget()->isVisibleTo(parentWidget()))
+        {
+            visibleItems.push_back(item);
+        }
+    }
+
+    auto firstOfRow = visibleItems.begin();
+    int lastOffset = -1;
+
+    for (auto iter = visibleItems.begin(); iter != visibleItems.end(); ++iter)
+    {
+        QLayoutItem* item = *iter;
+
+        QSize itemSize = item->sizeHint();
+        if (itemSize.width() > effectiveRect.width())
+        {
+            //if the single item is bigger than the row, scale it
+            if (itemSize.width() > 0)
+            {
+                itemSize.setWidth(effectiveRect.width());
+                if (item->hasHeightForWidth())
+                {
+                    //adjust the height if this item has an "height for width" layout
+                    itemSize.setHeight(item->heightForWidth(itemSize.width()));
+                }
+            }
+            itemSize.boundedTo(item->minimumSize());
+        }
+        QWidget* wid = item->widget();
+        int spaceX = horizontalSpacing();
+        if (spaceX == -1)
+        {
+            spaceX = wid->style()->layoutSpacing(
+                QSizePolicy::PushButton, QSizePolicy::PushButton, Qt::Horizontal);
+        }
+        int spaceY = verticalSpacing();
+        if (spaceY == -1)
+        {
+            spaceY = wid->style()->layoutSpacing(
+                QSizePolicy::PushButton, QSizePolicy::PushButton, Qt::Vertical);
+        }
+
+        int nextX = x + itemSize.width() + spaceX;
+        if (nextX - spaceX > effectiveRect.right() && lineHeight > 0)
+        {
+            if (!testOnly && (alignment() & Qt::AlignHCenter))
+            {
+                int w = (x - spaceX) - (*firstOfRow)->geometry().left();
```

Comment:
Nit: There are lots of variables in scope in this function, so it would be nice to use "const" when applicable.